### ISSUE TITLE
GrainClient.Uninitialize() synchronization issue, infinite 100% cpu loop

### ISSUE THREAD

**Issue by dolly22** [dotnet/orleans#3171]
I'm experiencing some multithreading issue with `GrainClient.Uninitialize()` resulting in infinite loop and 100% CPU core usage for the following scenario (version 1.4.2):

1. GrainClient is fully initialized and connected to silo
2. Thread 1 obtains some grain reference
3. Thread 2 calls `GrainClient.Uninitialize()`
4. After some small period of time Thread 1 makes grain call

```
var pingGrain = GrainClient.GrainFactory.GetGrain<IPingGrain>(0);
Task.Run(() => GrainClient.Uninitialize());
await Task.Delay(100);
await pingGrain.Echo("test");
```
When uninitializing `ProxiedMessageCenter.Stop()` changes `Running=false` and all active gateway connections are stopped. But when grain call from Thread 1 hits `ProxiedMessageCenter.SendMessage()` method, new `GatewayConnection `is created and started...

- `GatewayConnection.Start()` calls `Connect()` but does nothing as `MsgCenter.Running==false`
- `GatewayClientReceiver.Run()` loop calls `GatewayClientReceiver.FillBuffer()` which calls `GatewayConnection.Connect()` (because it's socket is null)
- `GatewayConnection.Connect()` just returns as `MsgCenter.Running==false`
- It seems the thread is now stuck in infinite loop and probably never recovers

<details>
  <summary>Relevant client log fragment</summary>

```
[2017-06-24 21:04:42.747 GMT     7	INFO   	100000	OutsideRuntimeClient	192.168.10.10:0]	OutsideRuntimeClient.Reset(): client Id *cli/46941c9d	
[2017-06-24 21:04:42.754 GMT     7	VERBOSE 	101408	sync.SafeTimerBase	192.168.10.10:0]	Disposing timer sync.SafeTimerBase	
[2017-06-24 21:04:42.785 GMT     7	VERBOSE 	100000	Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Stopped agent	
[2017-06-24 21:04:42.798 GMT     7	VERBOSE 	100000	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Stopped agent	
[2017-06-24 21:04:42.831 GMT     8	VERBOSE 	100920	Messaging.ProxiedMessageCenter	192.168.10.10:0]	Received Invalid Operation exception -- exiting. System.InvalidOperationException: The collection argument is empty and has been marked as complete with regards to additions.
   at System.Collections.Concurrent.BlockingCollection`1.Take()
   at Orleans.Messaging.ProxiedMessageCenter.WaitMessage(Categories type, CancellationToken ct)	
[2017-06-24 21:04:42.846 GMT     9	INFO   	100312	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Stop request processed	
[2017-06-24 21:04:42.848 GMT     9	INFO   	100328	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Stopping AsyncAgent Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0 that runs on managed thread 9	
[2017-06-24 21:04:42.867 GMT    10	INFO   	100328	Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Stopping AsyncAgent Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0 that runs on managed thread 10	
[2017-06-24 21:04:42.870 GMT     7	VERBOSE 	101408	sync.SafeTimerBase	192.168.10.10:0]	Disposing timer sync.SafeTimerBase	
[2017-06-24 21:04:42.871 GMT     4	VERBOSE 	101404	sync.SafeTimerBase	192.168.10.10:0]	Creating timer sync.SafeTimerBase with dueTime=00:30:00 period=-00:00:00.0010000	
[2017-06-24 21:04:42.871 GMT     7	VERBOSE 	101408	asynTask.SafeTimerBase	192.168.10.10:0]	Disposing timer asynTask.SafeTimerBase	
[2017-06-24 21:04:42.913 GMT     4	VERBOSE-2 	100000	OutsideRuntimeClient	192.168.10.10:0]	Send Request *cli/46941c9d@234995af->*grn/2E967722/00000000 #4: global::OrleansGrainInterfaces.IPingGrain:Echo()	
[2017-06-24 21:04:42.914 GMT     4	VERBOSE-2 	100917	Messaging.ProxiedMessageCenter	192.168.10.10:0]	Starting new bucket index 7466 for ordered messages to grain *grn/2E967722/00000000	
[2017-06-24 21:04:42.914 GMT     4	VERBOSE 	100916	Messaging.ProxiedMessageCenter	192.168.10.10:0]	Creating gateway to gwy.tcp://127.0.0.1:40000/0 for message to grain *grn/2E967722/00000000, bucket 7466, grain id hash code 560afd2aX	
[2017-06-24 21:04:42.915 GMT     4	VERBOSE 	100914	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Starting gateway connection for gateway gwy.tcp://127.0.0.1:40000/0	
[2017-06-24 21:04:42.915 GMT     4	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.916 GMT     4	VERBOSE 	100000	Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Started asynch agent Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0	
[2017-06-24 21:04:42.917 GMT    14	INFO   	100000	Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Starting AsyncAgent Messaging.GatewayClientReceiver/gwy.tcp://127.0.0.1:40000/0 on managed thread 14	
[2017-06-24 21:04:42.918 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.919 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.940 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.941 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.942 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.942 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.920 GMT    13	INFO   	100000	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Starting AsyncAgent Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0 on managed thread 13	
[2017-06-24 21:04:42.919 GMT     4	VERBOSE 	100000	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Started asynch agent Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	
[2017-06-24 21:04:42.945 GMT    13	VERBOSE-2 	100000	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Got a Request message to send: Request *cli/46941c9d@234995af->*grn/2E967722/00000000 #4: global::OrleansGrainInterfaces.IPingGrain:Echo()	
[2017-06-24 21:04:42.945 GMT     4	VERBOSE-2 	100918	Messaging.ProxiedMessageCenter	192.168.10.10:0]	Sending message Request *cli/46941c9d@234995af->*grn/2E967722/00000000 #4: global::OrleansGrainInterfaces.IPingGrain:Echo() via gateway gwy.tcp://127.0.0.1:40000/0	
[2017-06-24 21:04:42.966 GMT    13	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.943 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.989 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.989 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
[2017-06-24 21:04:42.991 GMT    14	VERBOSE 	100910	Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://127.0.0.1:40000/0	192.168.10.10:0]	Ignoring connection attempt to gateway gwy.tcp://127.0.0.1:40000/0 because the proxy message center is not running	
```

</details>

**Comment by gabikliot** [comment]
sounds like something that can be easily fixed by putting one of the components in the loop into Stopped state and checking it, such that we don't confuse "yet not started" with "already stopped" state.

**Comment by benjaminpetit** [comment]
It seems to be a very edge case since `GrainClient.Uninitialize` set `RuntimeClient.Current` to `null`. I couldn't repro this locally, instead I got a `NullReferenceException`.

But we could add some logic to avoid what you are experiencing yes

**Comment by sergeybykov** [comment]
Resolved via #4269.
